### PR TITLE
[SPARK-25525][SQL][PYSPARK] Do not update conf for existing SparkContext in SparkSession.getOrCreate.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

In [SPARK-20946](https://issues.apache.org/jira/browse/SPARK-20946), we modified `SparkSession.getOrCreate` to not update conf for existing `SparkContext` because `SparkContext` is shared by all sessions.
We should not update it in PySpark side as well.

## How was this patch tested?

Added tests.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan

File: python/pyspark/sql/context.py [hunk 1]
```diff
@@ -485,7 +485,8 @@ def __init__(self, sparkContext, jhiveContext=None):
             "SparkSession.builder.enableHiveSupport().getOrCreate() instead.",
             DeprecationWarning)
         if jhiveContext is None:
-            sparkSession = SparkSession.builder.enableHiveSupport().getOrCreate()
+            sparkContext._conf.set("spark.sql.catalogImplementation", "hive")
+            sparkSession = SparkSession.builder._sparkContext(sparkContext).getOrCreate()
```
Comment (cloud-fan) [inline python/pyspark/sql/context.py]: why this change?

**Review by dongjoon-hyun — COMMENTED** [review]
@ueshin . #22545 (this) is very similar with #22552 in terms of code.
Actually, #22552 looks like a subset of this. Given that, do we need both PRs?

**Comment by ueshin** [comment]
@dongjoon-hyun I'm  sorry that I haven't shared the context, but I talked with @cloud-fan about https://github.com/apache/spark/pull/22545#discussion_r220414608 off-line, and he wanted to separate the `HiveContext` fix into another pr, so I submitted #22552.

**Comment by cloud-fan** [comment]
LGTM

File: python/pyspark/sql/session.py [hunk 2]
```diff
@@ -181,17 +181,11 @@ def getOrCreate(self):
                             sparkConf.set(key, value)
                         sc = SparkContext.getOrCreate(sparkConf)
                         # This SparkContext may be an existing one.
```
Comment (HyukjinKwon) [inline python/pyspark/sql/session.py]: tiny nit: can we move this comment above `sc = ...`

**Comment by HyukjinKwon** [comment]
@cloud-fan, do we target this 2.4? Looks it might break an existing app, in particular, when a Python shell creates a session and another shell (like Zeppelin) or another session depends on a configuration in spark context. For instance, the fixed doctest:

```python
s1 = SparkSession.builder.config("k1", "v1").getOrCreate()
>>> s1 = SparkSession.builder.config("k1", "v1").getOrCreate()
>>> s1.conf.get("k1") == s1.sparkContext.getConf().get("k1") == "v1"
```

Looks like a bug but the behaviour change by this looks potentially quite crucial.

**Review by HyukjinKwon — APPROVED** [review]
LGTM apart from if we should backport or not

**Comment by cloud-fan** [comment]
The scala side change is already in 2.3.0. If we are ok with the behavior inconsistency between python and scala, it's fine to merge it to master only (and revert #22552 from 2.4 as well).

**Comment by HyukjinKwon** [comment]
I think the session support is kind of partially implemented in Python side, and not being very well tested. There are some inconsistency between Python and Scala side (for instance see https://github.com/apache/spark/pull/21990). I was also thinking of targeting this one only into master.

If the release is not quite close and the code here's well tested and implemented, I'd be happy with going to branch-2.4 but .. if you guys are fine, I hope we can target this only into master.

WDYT, @ueshin?

**Comment by HyukjinKwon** [comment]
Anyway, merged to master.

Let me take #22552 out of branch-2.4 for now but please feel free to get this and that into branch-2.4 (without checking along with me) if you guys feel strongly.

**Comment by ueshin** [comment]
I'm okay with merging this only into master.

**Comment by cloud-fan** [comment]
SGTM

File: python/pyspark/sql/session.py [hunk 3]
```diff
@@ -156,7 +156,7 @@ def getOrCreate(self):
             default.
 
             >>> s1 = SparkSession.builder.config("k1", "v1").getOrCreate()
-            >>> s1.conf.get("k1") == s1.sparkContext.getConf().get("k1") == "v1"
+            >>> s1.conf.get("k1") == "v1"
```
Comment (gatorsmile) [inline python/pyspark/sql/session.py]: @ueshin Could we also update the migration guide about this change?

Comment (HyukjinKwon) [reply python/pyspark/sql/session.py]: In that case, we might have to put the behaviour changes by https://github.com/apache/spark/pull/18536 together to the migration guide as well.

Comment (gatorsmile) [reply python/pyspark/sql/session.py]: We can do it together.

Comment (ueshin) [reply python/pyspark/sql/session.py]: Submitted a pr to update the migration guide #22682.
